Add Tekton task to configure sealights for Python. (#86)
konflux-ci/tekton-integration-catalog: docs/qe-available-tasks/sealights/python-instrumentation/python-instrumentation.md

# Sealights Python Instrumentation Tekton Task

> This Tekton task automates the process of instrumenting python code with Sealights for quality analytics and testing. It retrieves the source code from a trusted artifact, installs Python Sealights agent, configures the app for sealights using vars from your pipeline run, scans all .py files that are not in the --exclude step above and reports scan to sealights, and stores results to be used later on in testing. The task can be triggered by either Pull Request or other events.

Reviewer: ```suggestion
This Tekton task automates the process of instrumenting python code with Sealights for quality analytics and testing. It retrieves the source code from a trusted artifact, installs Python Sealights agent, configures the app for Sealights using vars from your PipelineRun, scans all .py files that are not in the --exclude step above and reports scan to Sealights, and stores results to be used later on in testing. The task can be triggered by either Pull Request or other events.
```

> This Tekton task automates the process of instrumenting python code with Sealights for quality analytics and testing. It retrieves the source code from a trusted artifact, installs Python Sealights agent, configures the app for sealights using vars from your pipeline run, scans all .py files that are not in the --exclude step above and reports scan to sealights, and stores results to be used later on in testing. The task can be triggered by either Pull Request or other events.

Reviewer: btw what do you mean by `--exclude step above` - I guess that's copy-pasted from the Tekton Task's script?

## Overview

This task performs the following steps:

1. **Retrieves** the source code from a trusted artifact.
2. **Configures & Scans** the Python application using Sealights.

The task can be triggered by different events (e.g., Pull Request, Push) and allows users to exclude specific Python packages from the configuration process.

## Parameters

| Name                  | Type     | Default       | Description                                                                                   |
|-----------------------|----------|---------------|-----------------------------------------------------------------------------------------------|
| `source-artifact`     | `string` | -             | The Trusted Artifact URI pointing to the source code.                                         |
| `python-version`          | `string` | -             | The Python version to use with the 'ubi8/python' image, in the format (e.g., '311').                                                       |

> | `sealights-secret`    | `string` | -             | The name of the OpenShift secret containing Sealights credentials.                            |

Reviewer: should we mention what secret key name is expected?

Author: hmm I'm now thinking that this var isn't actually needed, I'll need to check this.

Reviewer 2: hmm, I thought you need a token to send the data to Sealights when scanning

Author: Ya we do we are just getting the token from injecting a volume we don't need to pass it as a param like I had hear. I just tested the removal of this param and everything works fine.

| `component`           | `string` | -             | The name of the Konflux component associated with the integration tests.                      |
| `scm-provider`        | `string` | `github`      | The SCM provider (e.g., `github`, `gitlab`).                                                  |

> | `packages-excluded`   | `array`  | `[]`          | List of Go packages to exclude from instrumentation (e.g., `pkg1`, `github.com/lib/concurrent`). |

Reviewer: ```suggestion
| `packages-excluded`   | `array`  | `[]`          | List of Python packages to exclude from instrumentation (e.g., `pkg1`, `github.com/lib/concurrent`). |
```
and I'm not sure if the examples are also correct?

| `repository-url`      | `string` | `""`          | URL of the source code repository (e.g., `github.com/org/repo`).                              |
| `branch`              | `string` | `main`        | The Git branch to use (e.g., `main`, `feature-branch`).                                       |
| `revision`            | `string` | -             | The Git revision (commit SHA).                                                                |
| `pull-request-number` | `string` | `""`          | The Pull Request number.                                                                      |
| `target-branch`       | `string` | `main`        | The target branch for the Pull Request (e.g., `main`, `develop`).                             |
| `oci-storage`         | `string` | -             | The OCI repository for storing the trusted artifacts.                                         |

## Results

| Name                | Type     | Description                                                                 |
|---------------------|----------|-----------------------------------------------------------------------------|

> | `build-session-id`  | `string` | A unique identifier for the Sealights build session.                       |
> | `build-name`        | `string` | A unique build name generated using the commit SHA and current date.       |

Reviewer: ```suggestion
| `sealights-bsid`  | `string` | A unique identifier for the Sealights build session.                       |
| `sealights-build-name`        | `string` | A unique build name generated using the commit SHA and current date.       |
```

We assign the SEALIGHTS_TOKEN var in the script with this command.
```SEALIGHTS_TOKEN="$(cat /usr/local/sealights-credentials/token)"```
>NOTE: you must name the value of the secret **token**.

For example:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: sealights-credentials
type: Opaque
data:
  token: <BASE64_ENCODED_SEALIGHTS_TOKEN>
```

### Example Pipeline

Here's an example of how you can use the `sealights-python-instrumentation` task in a Tekton pipelinerun during Konflux CI build.

```yaml
    - name: sealights-instrumentation
      runAfter:
        - clone-repository
      taskRef:
        resolver: git
        params:
          - name: url
            value: https://github.com/konflux-ci/tekton-integration-catalog.git
          - name: revision
            value: main
          - name: pathInRepo
            value: tasks/sealights/python-instrumentation/0.1/python-instrumentation.yaml
      params:
        - name: source-artifact
          value: $(tasks.clone-repository.results.SOURCE_ARTIFACT)
        - name: python-version
          value: "311"
        - name: sealights-secret

>           value: "sealights-credentials"

Reviewer: I think this one should be removed now

| `branch`              | `string` | `main`        | The Git branch to use (e.g., `main`, `feature-branch`).                                       |
| `revision`            | `string` | -             | The Git revision (commit SHA).                                                                |
| `pull-request-number` | `string` | `""`          | The Pull Request number.                                                                      |
| `target-branch`       | `string` | `main`        | The target branch for the Pull Request (e.g., `main`, `develop`).                             |

## Results

| Name                | Type     | Description                                                                 |
|---------------------|----------|-----------------------------------------------------------------------------|
| `sealights-bsid`    | `string` | A unique identifier for the Sealights build session.                       |
| `sealights-build-name`        | `string` | A unique build name generated using the commit SHA and current date.       |

> | `source-artifact`   | `string` | The URI of the trusted artifact with the application source code.          |

Reviewer: we probably don't need this as a result for python?

Author: Hm your right I don't believe I use this outside of the python-instrumentation task which means I wouldn't need to store it as a result

## Overview

This task performs the following steps:

1. **Retrieves** the source code from a trusted artifact.
2. **Configures & Scans** the Python application using Sealights.

The task can be triggered by different events (e.g., Pull Request, Push) and allows users to exclude specific Python packages from the configuration process.

## Parameters

| Name                  | Type     | Default       | Description                                                                                   |
|-----------------------|----------|---------------|-----------------------------------------------------------------------------------------------|
| `source-artifact`     | `string` | -             | The Trusted Artifact URI pointing to the source code.                                         |
| `python-version`      | `string` | -             | The Python version to use with the 'ubi8/python' image, in the format (e.g., '311').          |                                             |
| `component`           | `string` | -             | The name of the Konflux component associated with the integration tests.                      |

> | `scm-provider`        | `string` | `git`         | The SCM provider (e.g., `git`).                                                               |

Author: Good catch yes I'll fix this

> | `packages-excluded`   | `array`  | `[]`          | A list of paths to exclude from Sealights instrumentation during the code scan. Specify paths to prevent them from being analyzed (e.g., '/app/source/tests/*','/app/examples/*'). |

Reviewer: So based on the description "A list of paths to exclude from Sealights instrumentation..." I'm wondering whether 
this should be then rather called `paths-excluded`?

Not strongly against the current param name, just wondering whether word "packages" might confuse user... If you think not, I'm fine with keeping it.

Reviewer 2: packages is more golang world, agree with paths or paths-to-exclude

Author: I had that at first, something like paths-excluded but then after experiencing the different cli's offered by sealights and how each one has the same option that is just named slightly different or uses a - instead of _ it got me thinking that I didn't want to do that to our users.

So I wanted to keep the packages param excluded exactly the same across all of our work ([same as go](https://github.com/konflux-ci/tekton-integration-catalog/blob/main/tasks/sealights/sealights-go-instrumentation/0.1/sealights-go-instrumentation.yaml#L46), then nodejs use the same param)

maybe this isn't a good enough reason :shrug: let me know what you think

Reviewer: I agree that having a different param names for Go vs. Python vs. Nodejs might cause a confusion.

So due to these differences in naming of this parameter for different languages, should we consider renaming it to `excluded` or `exclude` in all tasks? 🤷

Reviewer 2: i would like to have the same name like sealights cli flag..

Author: Each sealights tool names this differently

For the slcli it is named`--packages-excluded`
for sl-python it is named `--exclude`

I'd vote to rename all 3 to use `exclude` I can update mine here, as for using the same for nodejs and go. Its not a requirement just a nice thought. Fla is using the flag name that they use I think that is good enough too. So for nodejs we can just used whichever flag name they use too.

For example:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: sealights-credentials
type: Opaque
data:
  token: <BASE64_ENCODED_SEALIGHTS_TOKEN>
```

### Example Pipeline

Here's an example of how you can use the `sealights-python-instrumentation` task in a Tekton pipelinerun during Konflux CI build.

```yaml
    - name: sealights-instrumentation
      runAfter:
        - clone-repository
      taskRef:
        resolver: git
        params:
          - name: url
            value: https://github.com/konflux-ci/tekton-integration-catalog.git
          - name: revision
            value: main
          - name: pathInRepo
            value: tasks/sealights/python-instrumentation/0.1/python-instrumentation.yaml
      params:
        - name: source-artifact
          value: $(tasks.clone-repository.results.SOURCE_ARTIFACT)
        - name: python-version
          value: "311"
        - name: sealights-secret
          value: "sealights-credentials"
        - name: component
          value: '{{ repo_name }}'
        - name: branch
          value: '{{ source_branch }}'
        - name: revision
          value: '{{ revision }}'
        - name: oci-storage

>           value: $(params.output-image).sealights.git

Reviewer: we probably don't need this parameter anymore right

Author: Correct I'll remove it
